src/main/kotlin/com/trackforever/controllers/ProjectController.kt

```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
+                    val currentIssue = it
+                    lateinit var matchingIssue: TrackForeverIssue // Get the matching issue if there is one
+                    currentProject.issues.values.forEach { // Go through each issue and check prevHash matches
+                        if (it.hash == currentIssue.prevHash) { // there is a match
+                            matchingIssue = it // TODO: some optimization here
+                        }
+                    }
+                    try { // check if the match exists
+                        logger.debug("Preparing to update ${matchingIssue.id} from ${currentProject.id}")
+                        currentProject.issues.remove(matchingIssue.id)
+                    } catch (e: UninitializedPropertyAccessException) { // matchingIssue was never initialized
+                        // No need to delete anything.
+                    }
+                    val calculatedHash = generateHash(currentIssue)
+                    if (calculatedHash != currentIssue.hash) {
+                        logger.debug("Mismatched issue hash detected.")
+                        logger.debug("Given hash: ${currentIssue.hash}")
+                        logger.debug("Expected hash: $calculatedHash")
+                        currentIssue.hash = calculatedHash
+                    }
+                    val issuesList = successes[currentIssue.projectId]
+                    if (issuesList != null) {
+                        issuesList += currentIssue
+                    }
+                    currentProject.issues[currentIssue.id] = currentIssue
                 }
+                projectRepository.save(currentProject) // Update the entry in the database
             }
         }
         return when {
-            failedProjects.isEmpty() -> ResponseEntity(HttpStatus.OK)
-            failedProjects.keys.size == content.keys.size -> ResponseEntity(content, HttpStatus.GONE) // None of the specified projects exist.
-            else -> ResponseEntity(failedProjects, HttpStatus.MULTI_STATUS)
+            successes.isEmpty() -> ResponseEntity(HttpStatus.GONE) // None of the projects were found.
+            successes.keys.size == content.keys.size -> ResponseEntity(successes, HttpStatus.OK) // All projects were found
+            else -> ResponseEntity(successes, HttpStatus.MULTI_STATUS)
         }
     }
 
     /**
      * Adds the given projects to the database.
      * Expects a List of projects in a JSON String.
-     * Possible Responses: HTTP 200 OK upon completion.
-     * TODO: Verify that the projects being passed in are valid
+     * Possible Responses: HTTP 200 OK upon completion with a List of projects
+     *                     that were successfully saved.
      */
     @CrossOrigin
     @PutMapping("/projects")
-    fun setProjects(@RequestBody projects: String): ResponseEntity<String> {
+    fun setProjects(@RequestBody projects: String): ResponseEntity<List<TrackForeverProject>> {
```

Comment:
Return a List<ProjectId, Hash> here instead.

Reply:
A List or a Map here?

Reply:
Oh my bad. I mean Map<ProjectId, Hash>
Hash as in the project hash itself - a string.